sklearn/manifold/_isomap.py

```diff
@@ -186,9 +192,54 @@ def _fit_transform(self, X):
             n_jobs=self.n_jobs,
         )
 
-        self.dist_matrix_ = graph_shortest_path(
-            kng, method=self.path_method, directed=False
-        )
+        # Compute the number of connected components, and connect the different
+        # components to be able to compute a shortest path between all pairs
+        # of samples in the graph.
+        # Similar fix to cluster._agglomerative._fix_connectivity.
+        n_connected_components, labels = connected_components(kng)
+        if n_connected_components > 1:
+            if self.metric == "precomputed":
+                raise RuntimeError(
+                    "The number of connected components of the neighbors graph"
+                    f" is {n_connected_components} > 1. The graph cannot be "
+                    "completed with metric='precomputed', and Isomap cannot be"
+                    "fitted. Increase the number of neighbors to avoid this "
+                    "issue."
+                )
+            warnings.warn(
+                "The number of connected components of the neighbors graph "
+                f"is {n_connected_components} > 1. Completing the graph to fit"
+                " Isomap might be slow. Increase the number of neighbors to "
+                "avoid this issue.",
+                stacklevel=2,
+            )
+            # use array validated by NearestNeighbors
+            X = self.nbrs_._fit_X
+
+            for i in range(n_connected_components):
+                idx_i = np.where(labels == i)[0]
```

Comment:
Why not use np.nonzeros directly (here and everywhere)? As stated in [numpy docs](https://numpy.org/doc/stable/reference/generated/numpy.where.html),
```
When only condition is provided, this function is a shorthand for
np.asarray(condition).nonzero(). Using nonzero directly should be 
preferred, as it behaves correctly for subclasses. The rest of this
documentation covers only the case where all three arguments are provided.
```

Reply:
even `np.flatnonzero` then